.circleci/mkci

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python3
+"""Build application packages and upload them to bintray.
+Works on TravisCI, CircleCI and locally. Detect which package need to be rebuilt.
+Support "release" packages and PR/testing.
+"""
+
+# debdeps: git
+from argparse import ArgumentParser
+from os import getenv
+from pathlib import Path
+from requests.auth import HTTPBasicAuth
+from subprocess import run
+from typing import List
+import requests
+import sys
+
+BINTRAY_API = "https://bintray.com/api/v1"
+DEFAULT_ORG = "ooni"
+DEFAULT_REPO = "internal-pull-requests"
+
+
+def run2(cmd, **kw):
+    p = run(cmd.split(), capture_output=True, **kw)
+    if p.returncode != 0:
+        raise Exception(f"'{cmd}' returned: {p.returncode}")
+    return p.stdout.decode()
+
+
+def runi(cmd: str, cwd: Path, sudo=False) -> None:
+    if sudo:
+        cmd = f"sudo {cmd}"
+    run(cmd.split(), cwd=cwd, check=True)
+
+
+def detect_changed_packages() -> List[Path]:
+    DCH = "debian/changelog"
+    commit = run2("git merge-base master HEAD")
+    changes = run2(f"git diff --name-only {commit}")
+    changes += "af/fastpath\n"
+    pkgs = set()
+    for c in changes.splitlines():
+        c = Path(c)
+        if c.as_posix().endswith(DCH):
+            pkgs.add(c.parent.parent)
+            continue
+        while c.name:
+            if c.joinpath(DCH).is_file():
+                pkgs.add(c)
+            c = c.parent
+
+    return sorted(pkgs)
+
+
+def trim_compare(url: str) -> str:
+    """Shorten GitHub URLs used to compare changes"""
+    if url.startswith("https://github.com/") and "..." in url:
+        base, commits = url.rsplit("/", 1)
+        if len(commits) == 83:
+            beginning = commits[0:8]
+            end = commits[43:43+8]
+            return f"{base}/{beginning}...{end}"
+
+    return url
+
+
+
+def buildpkg(p) -> List[Path]:
+    """Build one package, installing required dependencies"""
+    print(f"Building package in {p}")
+    ver = run2("dpkg-parsechangelog --show-field Version", cwd=p).strip()
+    assert ver, f"No version number found in {p}/debian/changelog"
+    sudo = True
+    if getenv("CIRCLECI"):
+        # Running in CircleCI
+        sudo = False
+        comp = trim_compare(getenv("CIRCLE_COMPARE_URL", ""))  # show changes in VCS
+        if getenv("CIRCLE_PULL_REQUEST"):
+            # This is a PR: build ~pr<N>-<N> version. CIRCLE_PR_NUMBER is broken
+            pr_num = getenv("CIRCLE_PULL_REQUEST", "").rsplit("/", 1)[-1]
+            build_num = getenv("CIRCLE_BUILD_NUM")
+            ver = f"{ver}~pr{pr_num}-{build_num}"
```

Comment:
Maybe this is something which can go in a dedicated function or higher up in the script as a constant?

Reply:
I'm planning of moving the CircleCI and Bintray-specific code out in future to make the "core" service-independent.